### PR TITLE
[MINOR] Use standard java API to interrupt thread

### PR DESCRIPTION
**Pull request by DrIgor**
### What is this PR for?
Use java `Thread.interrupt` method to stop job progress polling thread.

Standard API is:
* proper synchronized 
* able to interrupt `Thread.sleep` with any interval

### What type of PR is it?
[Refactoring]

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

### CODE REVIEW

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/Job.java [hunk 1]
```diff
@@ -182,22 +182,16 @@ public void run() {
...
     } catch (Throwable e) {
       LOGGER.error("Job failed", e);
-      progressUpdator.terminate();
       this.exception = e;
       setResult(e.getMessage());
       errorMessage = getStack(e);
       dateFinished = new Date();
     } finally {
+      if (progressUpdator != null) {
+        progressUpdator.interrupt();
```
Comment (felixcheung) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/Job.java]: isn't this changing it to call interrupt every time, even when it has no error?

**Comment by felixcheung** [comment]
for a change like this, it would be useful to open a JIRA and to discuss why we are making this change

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/Job.java (hunk 1 above)
Comment (DrIgor) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/Job.java]: `terminate` was called in both cases: on line 185 in case of success and on line 195 in catch block. So it is better to place it in `finally` to handle both cases.

**Comment by DrIgor** [comment]
@felixcheung `terminate` flag is shared across several threads and is accessed without any synchronization.
And we don't need to implement custom interruption flag because java already has it — `Thread.interrupt`

Standard flag is even better, because it can interrupt `Thread.sleep`.

This PR doesn't change any logic, it just replaces `terminate` with `Thread.interrupted()` and provides proper handling of `InterruptedException`

**Comment by DrIgor** [comment]
Ping-ping)

Do you have other questions? Do you think it's better to proper synchronize `terminate` flag instead?

I reviewed code to understand how zeppelin core works and found several multi threading issues. I would like to help to fix them so I created this PR

**Comment by felixcheung** [comment]
From my understanding, setting a flag to let the thread gracefully runs to end, seems safer than calling `interrupt` on it

**Comment by DrIgor** [comment]
`interrupt` just sets `Thread.isInterrupted` flag and does not immediately stop the thread. Then thread can check this flag and decide if it needs to stop. So in fact, termination logic remains the same.

Maybe you mixed up `Thread.interrupt` and `Thread.stop` methods? `Thread.stop` terminates a thread without graceful shutdown, `Thread.interrupt` doesn't.

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/Job.java [hunk 2]
```diff
@@ -182,22 +182,16 @@ public void run() {
       this.exception = null;
       errorMessage = null;
       dateFinished = new Date();
-      progressUpdator.terminate();
-    } catch (NullPointerException e) {
```
Comment (felixcheung) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/Job.java]: why is this removed?

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/JobProgressPoller.java [hunk 3]
```diff
@@ -21,48 +21,45 @@
...
         Thread.sleep(intervalMs);
-      } catch (InterruptedException e) {
-        logger.error("Exception in JobProgressPoller while run Thread.sleep", e);
       }
-    }
-  }
-
-  public void terminate() {
-    terminate = true;
+    } catch (InterruptedException ignored) {}
```
Comment (felixcheung) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/JobProgressPoller.java]: should this be logged or it will "never" happen since we are not calling thread.stop()?

**Comment by felixcheung** [comment]
Ok I wasn't aware of the thread.interrupt semantic in JVM.

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/Job.java (hunk 2 above)
Comment (DrIgor) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/Job.java]: Both catch block were identical and `catch (Throwable e)` can handle `NullPointerException` too

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/JobProgressPoller.java (hunk 3 above)
Comment (DrIgor) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/JobProgressPoller.java]: `InterruptedException` will be thrown if we call `interrupt` during `Thread.sleep`. I don't log it because it's expected

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/Job.java (hunk 2 above)
Comment (felixcheung) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/Job.java]: ok, I guess we didn't do anything different there

**Comment by felixcheung** [comment]
ok LGTM.
any comment from other reviewers?

**Comment by Leemoonsoo** [comment]
Code change Looks good to me.

**Comment by felixcheung** [comment]
merging if no more comment
